### PR TITLE
100% test coverage for FastqRecord

### PR DESCRIPTION
**Pull request by akiezun**
### Description

Adding tests for `FastqRecord`. It now has 100% line coverage

### Checklist

- [X] Code compiles correctly
- [X] New tests covering changes and new functionality
- [X] All tests passing

### CODE REVIEW

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
+package htsjdk.samtools.fastq;
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: license

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertSame(fastqRecord.getBaseQualityString(), fastqRecordCopy.getBaseQualityString());
+        Assert.assertSame(fastqRecord.getBaseQualityHeader(), fastqRecordCopy.getBaseQualityHeader());
+    }
+
+    @Test
+    public void testEqualsWithNullSeq() {
+        final String seqHeaderPrefix = "header";
+        final String seqLine = "GATTACA";
+        final String qualHeaderPrefix = "";
+        final String qualLine = ";<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: I don't think it is valid for the quality line to have a different number of characters as the sequence line.

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertNotSame(fastqRecord, fastqRecordCopy);
+        Assert.assertSame(fastqRecord.getReadString(), fastqRecordCopy.getReadString());
+        Assert.assertSame(fastqRecord.getBaseQualityString(), fastqRecordCopy.getBaseQualityString());
+        Assert.assertSame(fastqRecord.getBaseQualityHeader(), fastqRecordCopy.getBaseQualityHeader());
+    }
+
+    @Test
+    public void testCopyWithNullSeq() {
+        final String seqHeaderPrefix = "header";
+        final String seqLine = null;
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: I think if `seqLine` is `null`, then `qualLine` should be required to be `null`.

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertEquals(fastqRecord.getReadHeader(), seqHeaderPrefix);
+        Assert.assertEquals(fastqRecord.getBaseQualityString(), qualLine);
+        Assert.assertEquals(fastqRecord.getReadString(), seqLine);
+        Assert.assertNotNull(fastqRecord.toString());//just check not nullness
+        Assert.assertNotEquals(fastqRecord, null);
+        Assert.assertNotEquals(null, fastqRecord);
+        Assert.assertEquals(fastqRecord, fastqRecord);
+        Assert.assertNotEquals(fastqRecord, "fred");
+        Assert.assertNotEquals("fred", fastqRecord);
+        Assert.assertEquals(fastqRecord.length(), seqLine.length());
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: assert quality line and sequence line are the same length?

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertNotEquals(fastqRecord2.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord1.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord2.hashCode(), fastqRecord2.hashCode());
+    }
+
+    @Test
+    public void testEqualsWithNullHeader() {
+        final String seqLine = "GATTACA";
+        final String qualHeaderPrefix = "";
+        final String qualLine = ";<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: same comment about lengths

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertNotEquals(fastqRecord1.hashCode(), fastqRecord2.hashCode());
+        Assert.assertNotEquals(fastqRecord2.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord1.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord2.hashCode(), fastqRecord2.hashCode());
+    }
+    @Test
+    public void testEqualsWithNullBaseQualityHeader() {
+        final String seqHeaderPrefix = "header";
+        final String seqLine = "GATTACA";
+        final String qualLine = ";<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: same comment about lengths and `null`

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        Assert.assertNotEquals(fastqRecord2.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord1.hashCode(), fastqRecord1.hashCode());
+        Assert.assertEquals(fastqRecord2.hashCode(), fastqRecord2.hashCode());
+    }
+
+    @Test
+    public void testEqualsWithNullQualLine() {
+        final String seqHeaderPrefix = "header";
+        final String seqLine = "GATTACA";
+        final String qualLine = ";<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
```
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: same comment about lengths

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 2 above)
Comment (akiezun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: the ctor does not blow up. Should it? if yes, that would change the current semantics. The same question for other comments: should we change or keep the sematics of `FastqRecord`?

Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: Various sites (ex. wikipedia, MAQ's page, and even here [1]) all say that the seq and qual fields should be the same length, so I would prefer a change in semantics.  `FastqReader` also enforces this prior to creating a `FastqRecord`.

[1] http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2847217/

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 1 above)
Comment (yfarjoun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: Really? Hardly any of the test files have licenses...is this a thing we want to start now? (I'm not sure why they don't have a license, but I've noticed it...)

Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: Crap, totally my fault for not remembering that this was a test file, just a habit.

Comment (akiezun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: ok, no license then

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 2 above)
Comment (akiezun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: the spec also indicated that both seq and quals are non-null and not empty. Is that a correct understanding?

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 3 above)
Comment (akiezun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: i think neither can be null or empty according to the spec

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 4 above)
Comment (akiezun) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: done

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 2 above)
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: I think so, having empty sequences makes no sense, and `nulls` seem wrong too.  I think we get to decide here.

File: src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java (hunk 3 above)
Comment (nh13) [inline src/tests/java/htsjdk/samtools/fastq/FastqRecordTest.java]: If we are going to follow the spec, then yes.

File: src/java/htsjdk/samtools/fastq/FastqRecord.java [hunk 8]
```diff
@@ -24,18 +24,23 @@
...
+import java.util.Objects;
 
 /**
  * Represents a fastq record, fairly literally, i.e. without any conversion.
  */
 public class FastqRecord implements Serializable {
     private static final long serialVersionUID = 1L;
     private final String seqHeaderPrefix;
-    private final String seqLine;
+    private final String seqLine;            //not null, not empty
```
Comment (akiezun) [inline src/java/htsjdk/samtools/fastq/FastqRecord.java]: according to the FASTQ spec that @nh13 pointed out, these are not null and not empty. So I enforce it here and use that fact to simplify a bunch of code in this class. Tests get updated accordingly

**Comment by akiezun** [comment]
@nh13 can you re-review?

File: src/java/htsjdk/samtools/fastq/FastqRecord.java (hunk 8 above)
Comment (tfenne) [inline src/java/htsjdk/samtools/fastq/FastqRecord.java]: Just commenting to note that, while I think this is good and the right thing to do, this is a breaking change that will effect many places in the ecosystem, and so should be clearly noted in release notes and a corresponding bump in version number made.

**Comment by akiezun** [comment]
1 sec

File: src/java/htsjdk/samtools/fastq/FastqRecord.java (hunk 8 above)
Comment (akiezun) [inline src/java/htsjdk/samtools/fastq/FastqRecord.java]: agreed, this is a breaking API change. Where do we make notes of such things?

Comment (tfenne) [inline src/java/htsjdk/samtools/fastq/FastqRecord.java]: I'm not sure...  Perhaps we should have an open ticket per release, or use the release prep features to keep a running list of things like this for the next release?  @jacarey what do you think?

**Comment by akiezun** [comment]
weird, there seems to be some transient travis problem - some FTP tests failing (?) what is that?

**Comment by lindenb** [comment]
```
private final String seqLine;            //not null, not empty
```

but some cleaning tools like cutadapt could remove the whole sequence. What happends when reading such fastq files with FastqReader ?

**Comment by lbergelson** [comment]
@akiezun There are a bunch of tests that hit a live ftp server.  They fail sometimes.  It's annoying.  I've created #560 to replace them with a mocked ftp server at some point.

**Comment by nh13** [comment]
@lindenb my personal preference would be for the authors of `cutadapt` to update the tools to output valid FASTQs instead of zero length reads.  Myself and others have certainly run into issues with the output of `cutadapt` when fed into aligners, especially RNA-seq aligners, that can't handle zero-length reads (and they should have to).

@akiezun I am assigning this back to you.  Looking more closely, `FastqReader` enforces this spec, but `FastqWriter` doesn't.  I think in Picard, checking that `SamToFastq` and `IlluminaBasecallsToFastq` do the right thing, in particular when trimming (adapter, quality, or other) may cause a zero-length read.

**Comment by akiezun** [comment]
@nh13  can you clarify what you want to see? Should I change this code or client code or something else? I'm not sure.

**Comment by akiezun** [comment]
I guess @nh13 may be too busy. Anyone else have an opinion? I'm fine with just dropping this PR it's too much hassle

**Comment by nh13** [comment]
I would like any FASTQ code to follow the spec rather than allow things like empty reads. Others should weigh in too as the code has been non-conforming for a while so an argument for not changing or for pragmatism could be made.

**Comment by yfarjoun** [comment]
@akiezun could you split this into two PR's:

1. increase test coverage
2. change behavior regarding empty reads

??

**Comment by akiezun** [comment]
@yfarjoun removed breaking changes from this PR. can you review?

**Comment by akiezun** [comment]
@yfarjoun good to merge?

**Comment by yfarjoun** [comment]
👍

**Pull request merged**